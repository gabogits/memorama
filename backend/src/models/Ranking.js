import { Schema, model } from "mongoose";

const RankingSchema = new Schema({
  name: {
    type: String,
    require: true,
    trim: true,
  },
  score: {
    type: Number,
    require: true,
    default: 0,
  },
  time: {
    type: String,
    require: true,
    default: 0,
  },
  avatar: {
    type: String,
    require: true,
  },
  created: {
    type: Date,
    default: Date.now(),
  },
});

export default model("Ranking", RankingSchema);
