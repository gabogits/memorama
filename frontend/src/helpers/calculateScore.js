export const calculateScore = (maxTime, time, scoreMax) => {
  if (isNaN(maxTime) || isNaN(time) || isNaN(scoreMax) || time >= maxTime) {
    return 0;
  }
  let perc = ((time / maxTime) * 100).toFixed(3);
  let score = parseInt(scoreMax * ((100 - perc) / 100));

  return score;
};
