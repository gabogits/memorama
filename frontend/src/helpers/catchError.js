export const catchError = (error) => {
  let errorMsg = "";
  if (error.response) {
    errorMsg = error.response.data?.msg;
    if (error.response?.data?.error) {
      errorMsg = error.response.data.error.msg;
    }
  } else {
    errorMsg = error.msg;
  }

  return errorMsg;
};
