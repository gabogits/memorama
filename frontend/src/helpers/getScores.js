import { catchError } from "./catchError";
import apiClient from "../helpers/apiClient";

export const getScores = async () => {
  try {
    const url = "/ranking";
    let response = await apiClient.get(url);

    return response.data;
  } catch (error) {
    return catchError(error);
  }
};
