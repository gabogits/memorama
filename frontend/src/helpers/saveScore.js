import { catchError } from "./catchError";
import apiClient from "../helpers/apiClient";
import formatTime from "./formatTimer";

export const saveScore = async (
  numberPlayers,
  endGame,
  score,
  dataPlayers,
  timer
) => {
  if (numberPlayers === 1 && endGame) {
    try {
      const url = "/ranking";

      const values = {
        name: dataPlayers[0].name,
        score,
        time: formatTime(timer),
        avatar: dataPlayers[0].avatar,
      };

      const response = await apiClient.post(url, values);
      console.log(response);
    } catch (error) {
      console.log(error);

      throw new Error(catchError(error));
    }
  }
};
