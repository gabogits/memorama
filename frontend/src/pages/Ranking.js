import React, { useEffect, useState } from "react";

import Layout from "../components/layout/Layout";
import Loader from "../components/templates/Loader";
import { getScores } from "../helpers/getScores";

const Ranking = () => {
  const [scores, setScores] = useState([]);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    const fetch = async () => {
      setLoading(true);
      const scores = await getScores();
      setScores(scores);
      setLoading(false);
    };
    fetch();
  }, []);

  return (
    <Layout>
      {loading && <Loader />}
      {scores.length > 0 && (
        <div className="ranking-container">
          {scores.map((item) => (
            <div className="list-block" key={item._id}>
              <div className="list-item">
                <div className="left-content">
                  <div className={`block-icon }`}>
                    <img
                      src={require(`../assets/images/avatar/${item.avatar}.png`)}
                      alt="avatar"
                    />
                  </div>
                  <div className="text-block">
                    <h4>{item.name}</h4>
                    <p>{item.time}</p>
                  </div>
                </div>
                <div className="right-content">
                  <h3>{item.score}</h3>
                </div>
              </div>
            </div>
          ))}
        </div>
      )}
    </Layout>
  );
};

export default Ranking;
