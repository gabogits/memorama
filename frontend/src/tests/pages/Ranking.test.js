import React from "react";
import { mount } from "enzyme";
import { Router } from "react-router-dom";
import GameContext from "../../context/GameContext";
import { act } from "@testing-library/react";
import { getScores } from "../../helpers/getScores";
import Ranking from "../../pages/Ranking";
jest.mock("../../helpers/getScores");

describe("Pruebas en <Ranking />", () => {
  let historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: jest.fn(),
    listen: jest.fn(),
    createHref: jest.fn(),
  };
  const contextValue = {
    dispatch: jest.fn(),
    state: {
      numberPlayers: 0,
      dataPlayers: [],
      arrayImages: [],
      category: null,
    },
  };
  const waitForComponentToPaint = async (wrapper) => {
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve));
      wrapper.update();
    });
  };
  beforeEach(() => {
    getScores.mockClear();
  });
  test("Debe de pintarse el listado del ranking", async () => {
    getScores.mockReturnValue([
      { _id: 1, avatar: "witch", name: "juan", time: "00:01:20", score: 1000 },
      { _id: 2, avatar: "dracula", name: "beto", time: "00:02:20", score: 800 },
    ]);
    const wrapper = mount(
      <GameContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Ranking />
        </Router>
      </GameContext.Provider>
    );
    expect(wrapper.find(".ranking-container").exists()).toBe(false);
    expect(wrapper.find("Loader").exists()).toBe(true);
    await waitForComponentToPaint(wrapper);
    expect(wrapper.find("Loader").exists()).toBe(false);
    expect(wrapper.find(".ranking-container").exists()).toBe(true);

    expect(
      wrapper.find(".list-item").at(0).find(".text-block h4").text().trim()
    ).toBe("juan");
    expect(
      wrapper.find(".list-item").at(0).find(".right-content h3").text().trim()
    ).toBe("1000");

    expect(
      wrapper.find(".list-item").at(1).find(".text-block h4").text().trim()
    ).toBe("beto");
    expect(
      wrapper.find(".list-item").at(1).find(".right-content h3").text().trim()
    ).toBe("800");
    expect(wrapper).toMatchSnapshot();
  });
});
