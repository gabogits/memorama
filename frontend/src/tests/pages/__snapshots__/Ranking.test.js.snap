// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Ranking /> Debe de pintarse el listado del ranking 1`] = `
<main>
  <header>
    <div
      className=" container container-header"
    >
      <div
        className="logo"
      >
        <figure>
          <a
            onClick={[Function]}
          >
            <span
              className="logo-desk"
            >
              MEMORAMA
            </span>
            <span
              className="logo-mobile"
            >
              M
            </span>
             
          </a>
        </figure>
      </div>
      <nav>
        <div
          className="nav-left"
        />
        <div
          className="nav-right"
        >
          <div
            className="nav-list"
          >
            <ul>
              <li>
                <button
                  onClick={[Function]}
                >
                  Inicio
                </button>
              </li>
              <li>
                <a
                  aria-current={null}
                  onClick={[Function]}
                >
                  Ranking
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
  </header>
  <div
    className="ranking-container"
  >
    <div
      className="list-block"
      key="1"
    >
      <div
        className="list-item"
      >
        <div
          className="left-content"
        >
          <div
            className="block-icon }"
          >
            <img
              alt="avatar"
              src="witch.png"
            />
          </div>
          <div
            className="text-block"
          >
            <h4>
              juan
            </h4>
            <p>
              00:01:20
            </p>
          </div>
        </div>
        <div
          className="right-content"
        >
          <h3>
            1000
          </h3>
        </div>
      </div>
    </div>
    <div
      className="list-block"
      key="2"
    >
      <div
        className="list-item"
      >
        <div
          className="left-content"
        >
          <div
            className="block-icon }"
          >
            <img
              alt="avatar"
              src="dracula.png"
            />
          </div>
          <div
            className="text-block"
          >
            <h4>
              beto
            </h4>
            <p>
              00:02:20
            </p>
          </div>
        </div>
        <div
          className="right-content"
        >
          <h3>
            800
          </h3>
        </div>
      </div>
    </div>
  </div>
</main>
`;
